IP filter accepts invalid setting but then fails during cluster state application

**Elasticsearch version** (`bin/elasticsearch --version`): observed in 7.11.2 & 7.12.0, but likely others too.

**Plugins installed**: Cloud

**JVM version** (`java -version`): Bundled

**OS version** (`uname -a` if on a Unix-like system): Cloud

**Description of the problem including expected versus actual behavior**:

It is possible to set an invalid value for `xpack.security.transport.filter.allow` via the Put Cluster Settings API, and the invalid value is persisted in the cluster state, but this raises an exception during cluster state application which causes the cluster to enter a loop in which the master node repeatedly fails, wins another election, attempts to apply the invalid value, and fails again. The cluster is effectively unavailable while in this state. If the invalid value is set as a persistent setting then this cannot even be resolved via a full cluster restart.

You can work around this by stopping all the master-eligible nodes, running `elasticsearch-node remove-settings xpack.security.transport.filter.allow` on each of them, and then starting them up again. Alternatively you can disable IP filtering by setting `xpack.security.transport.filter.enabled: false` and `xpack.security.http.filter.enabled: false` in `elasticsearch.yml`, which will allow the cluster to form, and then the persistent setting can be fixed.

Not really sure which area covers this, is it security or distrib? I've guessed distrib.

**Steps to reproduce**:

Execute this on an otherwise healthy cluster:

```
PUT /_cluster/settings
{
  "persistent": {
    "xpack.security.transport.filter.allow": "https://example.com/"
  }
}
```

Observe that the master node is now repeatedly failing.

**Provide logs (if relevant)**:

```
[2021-04-25T14:58:29,074][INFO ][o.e.c.s.ClusterSettings  ] [node-1] updating [xpack.security.transport.filter.allow] from [[]] to [["https://example.com"]]
[2021-04-25T14:58:29,079][WARN ][o.e.c.s.ClusterSettings  ] [node-1] failed to apply settings
org.elasticsearch.ElasticsearchException: unable to create ip filter for rule [allow  https://example.com]
        at org.elasticsearch.xpack.security.transport.filter.SecurityIpFilterRule.getRule(SecurityIpFilterRule.java:174) ~[?:?]
        at org.elasticsearch.xpack.security.transport.filter.SecurityIpFilterRule.<init>(SecurityIpFilterRule.java:63) ~[?:?]
        at org.elasticsearch.xpack.security.transport.filter.IPFilter.createRules(IPFilter.java:277) ~[?:?]
        at org.elasticsearch.xpack.security.transport.filter.IPFilter.parseSettings(IPFilter.java:245) ~[?:?]
        at org.elasticsearch.xpack.security.transport.filter.IPFilter.updateRules(IPFilter.java:232) ~[?:?]
        at org.elasticsearch.xpack.security.transport.filter.IPFilter.setTransportAllowFilter(IPFilter.java:190) ~[?:?]
        at org.elasticsearch.common.settings.Setting$Updater.apply(Setting.java:1160) ~[elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.common.settings.AbstractScopedSettings$SettingUpdater.lambda$updater$0(AbstractScopedSettings.java:618) ~[elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.common.settings.AbstractScopedSettings.applySettings(AbstractScopedSettings.java:182) [elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:455) [elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:407) [elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.cluster.service.ClusterApplierService.access$000(ClusterApplierService.java:57) [elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:151) [elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:673) [elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:241) [elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:204) [elasticsearch-7.12.0.jar:7.12.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
        at java.lang.Thread.run(Thread.java:831) [?:?]
Caused by: java.net.UnknownHostException: https:: nodename nor servname provided, or not known
        at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method) ~[?:?]
        at java.net.InetAddress$PlatformNameService.lookupAllHostAddr(InetAddress.java:932) ~[?:?]
        at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1517) ~[?:?]
        at java.net.InetAddress$NameServiceAddresses.get(InetAddress.java:851) ~[?:?]
        at java.net.InetAddress.getAllByName0(InetAddress.java:1507) ~[?:?]
        at java.net.InetAddress.getAllByName(InetAddress.java:1366) ~[?:?]
        at java.net.InetAddress.getAllByName(InetAddress.java:1300) ~[?:?]
        at java.net.InetAddress.getByName(InetAddress.java:1250) ~[?:?]
        at org.elasticsearch.xpack.security.transport.filter.SecurityIpFilterRule.parseSubnetMask(SecurityIpFilterRule.java:95) ~[?:?]
        at org.elasticsearch.xpack.security.transport.filter.SecurityIpFilterRule.getRule(SecurityIpFilterRule.java:170) ~[?:?]
        ... 18 more
[2021-04-25T14:58:29,089][WARN ][o.e.c.s.ClusterApplierService] [node-1] failed to apply updated cluster state in [200ms]:
version [806], uuid [MQYUJMC2TYO119m74nLjgg], source [Publication{term=879, version=806}]
org.elasticsearch.ElasticsearchException: unable to create ip filter for rule [allow  https://example.com]
        at org.elasticsearch.xpack.security.transport.filter.SecurityIpFilterRule.getRule(SecurityIpFilterRule.java:174) ~[?:?]
        at org.elasticsearch.xpack.security.transport.filter.SecurityIpFilterRule.<init>(SecurityIpFilterRule.java:63) ~[?:?]
        at org.elasticsearch.xpack.security.transport.filter.IPFilter.createRules(IPFilter.java:277) ~[?:?]
        at org.elasticsearch.xpack.security.transport.filter.IPFilter.parseSettings(IPFilter.java:245) ~[?:?]
        at org.elasticsearch.xpack.security.transport.filter.IPFilter.updateRules(IPFilter.java:232) ~[?:?]
        at org.elasticsearch.xpack.security.transport.filter.IPFilter.setTransportAllowFilter(IPFilter.java:190) ~[?:?]
        at org.elasticsearch.common.settings.Setting$Updater.apply(Setting.java:1160) ~[elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.common.settings.AbstractScopedSettings$SettingUpdater.lambda$updater$0(AbstractScopedSettings.java:618) ~[elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.common.settings.AbstractScopedSettings.applySettings(AbstractScopedSettings.java:182) ~[elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:455) ~[elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:407) [elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.cluster.service.ClusterApplierService.access$000(ClusterApplierService.java:57) [elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:151) [elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:673) [elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:241) [elasticsearch-7.12.0.jar:7.12.0]
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:204) [elasticsearch-7.12.0.jar:7.12.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
        at java.lang.Thread.run(Thread.java:831) [?:?]
Caused by: java.net.UnknownHostException: https:: nodename nor servname provided, or not known
        at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method) ~[?:?]
        at java.net.InetAddress$PlatformNameService.lookupAllHostAddr(InetAddress.java:932) ~[?:?]
        at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1517) ~[?:?]
        at java.net.InetAddress$NameServiceAddresses.get(InetAddress.java:851) ~[?:?]
        at java.net.InetAddress.getAllByName0(InetAddress.java:1507) ~[?:?]
        at java.net.InetAddress.getAllByName(InetAddress.java:1366) ~[?:?]
        at java.net.InetAddress.getAllByName(InetAddress.java:1300) ~[?:?]
        at java.net.InetAddress.getByName(InetAddress.java:1250) ~[?:?]
        at org.elasticsearch.xpack.security.transport.filter.SecurityIpFilterRule.parseSubnetMask(SecurityIpFilterRule.java:95) ~[?:?]
        at org.elasticsearch.xpack.security.transport.filter.SecurityIpFilterRule.getRule(SecurityIpFilterRule.java:170) ~[?:?]
        ... 18 more
[2021-04-25T14:58:29,098][INFO ][o.e.c.s.ClusterApplierService] [node-1] master node changed {previous [{node-1}{UEwM9WBVQxOfK7Vrk8HvmA}{YraW7amCTQGGJ5qInTtidw}{127.0.0.1}{127.0.0.1:9301}{cdfhilmrstw}{ml.machine_memory=68719476736, xpack.installed=true, transform.node=true, ml.max_open_jobs=20, ml.max_jvm_s
ize=1073741824}], current []}, term: 879, version: 806, reason: becoming candidate: clusterApplier#onNewClusterState
[2021-04-25T14:58:29,110][INFO ][o.e.c.s.ClusterSettings  ] [node-1] updating [xpack.security.transport.filter.allow] from [[]] to [["https://example.com"]]
```

--- a/x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/transport/filter/IpFilteringUpdateTests.java
+++ b/x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/transport/filter/IpFilteringUpdateTests.java
@@ -20,6 +20,8 @@ import java.util.Locale;
 
 import static org.elasticsearch.test.ESIntegTestCase.Scope.TEST;
 import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.is;
 
 @ClusterScope(scope = TEST, supportsDedicatedMasters = false, numDataNodes = 1)
@@ -119,6 +121,41 @@ public class IpFilteringUpdateTests extends SecurityIntegTestCase {
             assertAcked(client().admin().cluster().prepareUpdateSettings().setPersistentSettings(settings));
             assertAcked(client().admin().cluster().prepareUpdateSettings().setTransientSettings(settings));
             assertConnectionAccepted(".http", "127.0.0.8");
+        }
+    }
+
+    public void testThatInvalidDynamicIpFilterConfigurationIsRejected() {
+        final Settings.Builder initialSettingsBuilder = Settings.builder();
+        if (randomBoolean()) {
+            initialSettingsBuilder.put(IPFilter.IP_FILTER_ENABLED_SETTING.getKey(), randomBoolean());
+        }
+        if (randomBoolean()) {
+            initialSettingsBuilder.put(IPFilter.IP_FILTER_ENABLED_HTTP_SETTING.getKey(), randomBoolean());
+        }
+        final Settings initialSettings = initialSettingsBuilder.build();
+        if (initialSettings.isEmpty() == false) {
+            updateSettings(initialSettings);
+        }
+
+        final String invalidValue = "http://";
+
+        for (final String settingPrefix : new String[] {
+                "xpack.security.transport.filter",
+                "xpack.security.http.filter",
+                "transport.profiles.default.xpack.security.filter",
+                "transport.profiles.anotherprofile.xpack.security.filter"
+        }) {
+            for (final String settingSuffix : new String[]{"allow", "deny"}) {
+                final String settingName = settingPrefix + "." + settingSuffix;
+                final Settings settings = Settings.builder().put(settingName, invalidValue).build();
+                assertThat(
+                        settingName,
+                        expectThrows(
+                                IllegalArgumentException.class,
+                                settingName,
+                                () -> updateSettings(settings)).getMessage(),
+                        allOf(containsString("invalid IP filter"), containsString(invalidValue)));
+            }
         }
     }
 

--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/filter/IPFilter.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/filter/IPFilter.java
@@ -98,14 +98,14 @@ public class IPFilter {
             "transport.profiles.default.xpack.security.filter.allow",
             TRANSPORT_FILTER_ALLOW_SETTING,
             Function.identity(),
-            s -> Collections.emptyList(),
+            TRANSPORT_FILTER_ALLOW_SETTING::get,
             ALLOW_VALIDATOR,
             Property.NodeScope);
     public static final Setting<List<String>> HTTP_FILTER_ALLOW_SETTING = Setting.listSetting(
             setting("http.filter.allow"),
             HTTP_FILTER_ALLOW_FALLBACK,
             Function.identity(),
-            s -> Collections.emptyList(),
+            HTTP_FILTER_ALLOW_FALLBACK::get,
             ALLOW_VALIDATOR,
             Property.OperatorDynamic,
             Property.NodeScope);
@@ -114,14 +114,14 @@ public class IPFilter {
             "transport.profiles.default.xpack.security.filter.deny",
             TRANSPORT_FILTER_DENY_SETTING,
             Function.identity(),
-            s -> Collections.emptyList(),
+            TRANSPORT_FILTER_DENY_SETTING::get,
             DENY_VALIDATOR,
             Property.NodeScope);
     public static final Setting<List<String>> HTTP_FILTER_DENY_SETTING = Setting.listSetting(
             setting("http.filter.deny"),
             HTTP_FILTER_DENY_FALLBACK,
             Function.identity(),
-            s -> Collections.emptyList(),
+            HTTP_FILTER_DENY_FALLBACK::get,
             DENY_VALIDATOR,
             Property.OperatorDynamic,
             Property.NodeScope);

--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/filter/IPFilter.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/filter/IPFilter.java
@@ -55,30 +55,76 @@ public class IPFilter {
     public static final Setting<Boolean> IP_FILTER_ENABLED_SETTING = Setting.boolSetting(setting("transport.filter.enabled"),
             true, Property.OperatorDynamic, Property.NodeScope);
 
-    public static final Setting<List<String>> TRANSPORT_FILTER_ALLOW_SETTING = Setting.listSetting(setting("transport.filter.allow"),
-            Collections.emptyList(), Function.identity(), Property.OperatorDynamic, Property.NodeScope);
+    private static final IPFilterValidator ALLOW_VALIDATOR = new IPFilterValidator(true);
+    private static final IPFilterValidator DENY_VALIDATOR = new IPFilterValidator(false);
 
-    public static final Setting<List<String>> TRANSPORT_FILTER_DENY_SETTING = Setting.listSetting(setting("transport.filter.deny"),
-            Collections.emptyList(), Function.identity(), Property.OperatorDynamic, Property.NodeScope);
+    public static final Setting<List<String>> TRANSPORT_FILTER_ALLOW_SETTING = Setting.listSetting(
+            setting("transport.filter.allow"),
+            Collections.emptyList(),
+            Function.identity(),
+            ALLOW_VALIDATOR,
+            Property.OperatorDynamic,
+            Property.NodeScope);
+    public static final Setting<List<String>> TRANSPORT_FILTER_DENY_SETTING = Setting.listSetting(
+            setting("transport.filter.deny"),
+            Collections.emptyList(),
+            Function.identity(),
+            DENY_VALIDATOR,
+            Property.OperatorDynamic,
+            Property.NodeScope);
 
-    public static final Setting.AffixSetting<List<String>> PROFILE_FILTER_DENY_SETTING = Setting.affixKeySetting("transport.profiles.",
-            "xpack.security.filter.deny", key -> Setting.listSetting(key, Collections.emptyList(), Function.identity(),
-            Property.OperatorDynamic, Property.NodeScope));
-    public static final Setting.AffixSetting<List<String>> PROFILE_FILTER_ALLOW_SETTING = Setting.affixKeySetting("transport.profiles.",
-            "xpack.security.filter.allow", key -> Setting.listSetting(key, Collections.emptyList(), Function.identity(),
-            Property.OperatorDynamic, Property.NodeScope));
+    public static final Setting.AffixSetting<List<String>> PROFILE_FILTER_DENY_SETTING = Setting.affixKeySetting(
+            "transport.profiles.",
+            "xpack.security.filter.deny",
+            key -> Setting.listSetting(
+                    key,
+                    Collections.emptyList(),
+                    Function.identity(),
+                    DENY_VALIDATOR,
+                    Property.OperatorDynamic,
+                    Property.NodeScope));
+    public static final Setting.AffixSetting<List<String>> PROFILE_FILTER_ALLOW_SETTING = Setting.affixKeySetting(
+            "transport.profiles.",
+            "xpack.security.filter.allow",
+            key -> Setting.listSetting(
+                    key,
+                    Collections.emptyList(),
+                    Function.identity(),
+                    ALLOW_VALIDATOR,
+                    Property.OperatorDynamic,
+                    Property.NodeScope));
 
-    private static final Setting<List<String>> HTTP_FILTER_ALLOW_FALLBACK =
-            Setting.listSetting("transport.profiles.default.xpack.security.filter.allow", TRANSPORT_FILTER_ALLOW_SETTING, s -> s,
-                    Property.NodeScope);
-    public static final Setting<List<String>> HTTP_FILTER_ALLOW_SETTING = Setting.listSetting(setting("http.filter.allow"),
-            HTTP_FILTER_ALLOW_FALLBACK, Function.identity(), Property.OperatorDynamic, Property.NodeScope);
+    private static final Setting<List<String>> HTTP_FILTER_ALLOW_FALLBACK = Setting.listSetting(
+            "transport.profiles.default.xpack.security.filter.allow",
+            TRANSPORT_FILTER_ALLOW_SETTING,
+            Function.identity(),
+            s -> Collections.emptyList(),
+            ALLOW_VALIDATOR,
+            Property.NodeScope);
+    public static final Setting<List<String>> HTTP_FILTER_ALLOW_SETTING = Setting.listSetting(
+            setting("http.filter.allow"),
+            HTTP_FILTER_ALLOW_FALLBACK,
+            Function.identity(),
+            s -> Collections.emptyList(),
+            ALLOW_VALIDATOR,
+            Property.OperatorDynamic,
+            Property.NodeScope);
 
-    private static final Setting<List<String>> HTTP_FILTER_DENY_FALLBACK =
-            Setting.listSetting("transport.profiles.default.xpack.security.filter.deny", TRANSPORT_FILTER_DENY_SETTING, s -> s,
-                    Property.NodeScope);
-    public static final Setting<List<String>> HTTP_FILTER_DENY_SETTING = Setting.listSetting(setting("http.filter.deny"),
-            HTTP_FILTER_DENY_FALLBACK, Function.identity(), Property.OperatorDynamic, Property.NodeScope);
+    private static final Setting<List<String>> HTTP_FILTER_DENY_FALLBACK = Setting.listSetting(
+            "transport.profiles.default.xpack.security.filter.deny",
+            TRANSPORT_FILTER_DENY_SETTING,
+            Function.identity(),
+            s -> Collections.emptyList(),
+            DENY_VALIDATOR,
+            Property.NodeScope);
+    public static final Setting<List<String>> HTTP_FILTER_DENY_SETTING = Setting.listSetting(
+            setting("http.filter.deny"),
+            HTTP_FILTER_DENY_FALLBACK,
+            Function.identity(),
+            s -> Collections.emptyList(),
+            DENY_VALIDATOR,
+            Property.OperatorDynamic,
+            Property.NodeScope);
 
     public static final Map<String, Object> DISABLED_USAGE_STATS = Map.of(
             "http", false,
@@ -304,4 +350,26 @@ public class IPFilter {
         settings.add(PROFILE_FILTER_ALLOW_SETTING);
         settings.add(PROFILE_FILTER_DENY_SETTING);
     }
+
+    private static class IPFilterValidator implements Setting.Validator<List<String>> {
+
+        private final boolean isAllowRule;
+
+        IPFilterValidator(boolean isAllowRule) {
+            this.isAllowRule = isAllowRule;
+        }
+
+        @Override
+        public void validate(List<String> filterSpecs) {
+            for (String filterSpec : filterSpecs) {
+                try {
+                    // It's enough just to create the specified rule:
+                    new SecurityIpFilterRule(isAllowRule, filterSpec);
+                } catch (Exception e) {
+                    throw new IllegalArgumentException("invalid IP filter [" + filterSpec + "]", e);
+                }
+            }
+        }
+    }
+
 }

